bitcoin/script/miniscript.h

```diff
@@ -1014,224 +1014,404 @@ bool DecomposeScript(const CScript& script, std::vector<std::pair<opcodetype, st
-        if (!sub) return {};
-        if (in == last || in[0].first != OP_TOALTSTACK) return {};
-        ++in;
-        return MakeNodeRef<Key>(NodeType::WRAP_A, Vector(std::move(sub)));
+            // could be andor or or_i
+            if (in[0].first == OP_ELSE) {
+                ++in;
+                to_parse.emplace_back(DecodeContext::ENDIF_ELSE, -1, -1);
+                to_parse.emplace_back(DecodeContext::BKV_EXPR, -1, -1);
+            }
+            // could be j: or d: wrapper
+            else if (in[0].first == OP_IF) {
+                if (last - in >= 2 && in[1].first == OP_DUP) {
+                    in += 2;
+                    to_parse.emplace_back(DecodeContext::DUP_IF, -1, -1);
+                } else if (last - in >= 3 && in[1].first == OP_0NOTEQUAL && in[2].first == OP_SIZE) {
+                    in += 3;
+                    to_parse.emplace_back(DecodeContext::NON_ZERO, -1, -1);
+                }
+                else {
+                    return {};
+                }
+            // could be or_c or or_d
+            } else if (in[0].first == OP_NOTIF) {
+                ++in;
+                to_parse.emplace_back(DecodeContext::ENDIF_NOTIF, -1, -1);
+            }
+            else {
+                return {};
+            }
+            break;
+        }
+        case DecodeContext::ENDIF_NOTIF: {
+            if (in >= last) return {};
+            if (in[0].first == OP_IFDUP) {
+                ++in;
+                to_parse.emplace_back(DecodeContext::OR_D, -1, -1);
+            } else {
+                to_parse.emplace_back(DecodeContext::OR_C, -1, -1);
+            }
+            // or_c and or_d both require X to have type modifier d so, can't contain and_v
+            to_parse.emplace_back(DecodeContext::SINGLE_BKV_EXPR, -1, -1);
+            break;
+        }
+        case DecodeContext::ENDIF_ELSE: {
+            if (in >= last) return {};
+            if (in[0].first == OP_IF) {
+                ++in;
+                NodeRef<Key> left = std::move(constructed.back());
+                constructed.pop_back();
+                NodeRef<Key> right = std::move(constructed.back());
+                constructed.back() = MakeNodeRef<Key>(NodeType::OR_I, Vector(std::move(left), std::move(right)));
+            } else if (in[0].first == OP_NOTIF) {
+                ++in;
+                to_parse.emplace_back(DecodeContext::ANDOR, -1, -1);
+                // andor requires X to have type modifier d, so it can't be and_v
+                to_parse.emplace_back(DecodeContext::SINGLE_BKV_EXPR, -1, -1);
+            } else {
+                return {};
+            }
+            break;
+        }
+        }
     }
-    auto sub = DecodeMulti<Key>(in, last, ctx);
-    if (!sub) return {};
-    if (in == last || in[0].first != OP_SWAP) return {};
-    ++in;
-    return MakeNodeRef<Key>(NodeType::WRAP_S, Vector(std::move(sub)));
+    if (constructed.size() != 1) return {};
```

Comment:
No need to change anything in the code here, but just a question: I think that `constructed.size() != 1` at this point would indicate a bug in the code. `DecodeContext::BKV_EXPR` should translate (if successful) to exactly 1 stack element.

Reply:
Yes, I think this can be replaced with assert.